DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/compute_units/src/compute_units.hpp

```diff
@@ -0,0 +1,37 @@
+//==============================================================
+// Copyright Intel Corporation
+//
+// SPDX-License-Identifier: MIT
+// =============================================================
+#include <CL/sycl.hpp>
+#include <utility>
+
+namespace {
+template <class Func, template <std::size_t> class Name, std::size_t index>
+class SubmitOneComputeUnit {
+public:
+  SubmitOneComputeUnit(Func &&f, sycl::queue &q) {
+
+    q.submit([&](sycl::handler &h) {
+      h.single_task<Name<index>>([=] {
+        // verifies that f only takes a single argument
```

Comment:
Yes, it is the intended comment.

The submit_compute_units function takes a lambda function (representing the kernel that should be duplicated) as an argument. That lambda function needs to satisfy certain properties. One of those properties is that the lambda function should take exactly one argument (the compute unit ID). If not, then the compile will fail at this line.

Reply:
I agree with Ronan here. Should the wording be "verifies that f only...". "verifies" is the weird part to me. It just won't compile. These static_asserts may be useful though (https://en.cppreference.com/w/cpp/types/is_invocable). Maybe something like:
```c++
static_assert(std::is_invocable<Func, std::integral_constant<std::size_t, index>>::value, "Your Error Message");
```
This will verify that Func is callable with a single `std::integral_constant<std::size_t, index>` operand.

Reply:
Perhaps `// Assume that Func is a callable with a std::size_t argument`?
But I would put this in the comment above describing `SubmitOneComputeUnit`

Reply:
Scanning the usage of `submit_compute_units` in submit_compute_units.cpp, this operand is used to index the `PipeArray`. There, it HAS to be a constant, hence the `std::integral_constant`

Reply:
Yes, that's right Tanner. The main challenge in writing the compute units API was needing to have a constexpr function argument. The user writes a lambda function containing their kernel's functionality. Within that kernel code, the user should be able to access the compute unit ID as a constexpr.

This was the starting point I found: https://stackoverflow.com/questions/56937863/constexpr-lambda-argument

I will add the static_assert, thanks for the suggestion!